Removing Cry Dependency for Getting the Current Level Name

AMZN-Gene opened the pull request:
- Adding GetCurrentLevelName() method to AzFramework LevelSystem in order to remove Cry dependency. 
- Multiplayer will need to use this value in an upcoming PR
- Rename level blocking with its own EBus

Signed-off-by: Gene Walters <genewalt@amazon.com>

_Please add links to any issues, RFCs or other items that are relevant to this PR._
A piece of #9666 

## How was this PR tested?
Making sure the updated code that was using the old GetLevelName is still working.

lumberyard-employee-dm commented on Code/Legacy/CrySystem/LevelSystem/LevelSystem.h:
```diff
@@ -92,6 +94,16 @@ class CLevelSystem
+        {
+            return m_pCurrentLevel->GetLevelInfo()->GetName();
+        }
+        return "";
```
returning "" is a pessimization.
A default constructed AZStd::string should be returned instead
```suggestion
        return AZStd::string{};
```

AMZN-Gene replied:
Using const char* thanks :)

lumberyard-employee-dm commented on Code/Framework/AzFramework/AzFramework/API/ApplicationAPI.h:
```diff
@@ -189,18 +189,32 @@ namespace AzFramework
+        static const AZ::EBusHandlerPolicy HandlerPolicy = AZ::EBusHandlerPolicy::Single;
+        static const AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::Single;
+
+        //! Returns the name of the currently loaded level.
+        //! Note: for spawnable level system, this is the cache folder path to the level asset. Example: levels/mylevel/mylevel.spawnable
+        //! @return Level name or empty string if no level loaded.
+        virtual AZStd::string GetCurrentLevelName() = 0;
```
This function should be const.
But honestly, this shouldn't even be an EBus.

It should be an interface that gets used with AZ::Interface.

Also this function can return a `const char*`
No need for a `string` at this moment.

```
class ILevelSystemLifecycle
{
    virtual const char* GetCurrentLevelName() const = 0;
};

using LevelSystemLifecycleInterface = AZ::Interface<ILevelSystemLifecycle>;
```

AMZN-Gene replied:
Much better :D
Side question: is there still rule against using const EBus functions? I recall that being a thing even for simple stuff like Getters().

lumberyard-employee-dm replied:
No, that paradigm of avoiding using const EBus functions was never built on a solid reason.

The thought process by those recommending it was what "if the handler function need to modify the class, but the interface imposed a const function it".

It is not a strong reason why functions shouldn't be able to marked const

AMZN-Gene replied:
Thank you. I'm a const fan.

lumberyard-employee-dm commented on Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp:
```diff
@@ -82,6 +82,7 @@ namespace LegacyLevelSystem
         }
 
         AzFramework::RootSpawnableNotificationBus::Handler::BusConnect();
+        AzFramework::LevelSystemLifecycleRequestBus::Handler::BusConnect();
```
This should instead register itself with an AZ::Interface.
```suggestion
        auto registerOutcome = AzFramework::LevelSystemLifecycleInterface::Register(this);
        AZ_Error("SpawnableLevelSystem", registerOutcome, "Failed to register the SpawnableLevelSystem with the LevelSystemLifecycleInterface: %s", registerOutcome.GetError().c_str());
```

AMZN-Gene replied:
Fixed using Interface now! Deriving from Interface<>::Registrar seems fine too, or should I explicitly keep the error checking?

lumberyard-employee-dm replied:
I think either work fine, though I would prefer the error message path.

AMZN-Gene replied:
Good idea. Keep the RAII, and the error.

pollend commented on Code/Legacy/CryCommon/ILevelSystem.h:
```diff
@@ -65,7 +65,11 @@ struct ILevelSystem
     virtual bool LoadLevel(const char* levelName) = 0;
     virtual void UnloadLevel() = 0;
-    virtual bool IsLevelLoaded() = 0;
+
+    // Deprecated. Use AzFramework::LevelSystemLifecycleInterface::Get()->IsLevelLoaded instead.
+    virtual bool IsLevelLoaded() const = 0;
+
+    // Deprecated. Use AzFramework::LevelSystemLifecycleInterface::Get()->GetCurrentLevelName instead.
```
can you follow this: https://github.com/o3de/sig-core/blob/main/governance/API-Ref-Guidelines-Update.md#deprecations

AMZN-Gene replied:
fixed! thank you

pollend commented:
@hultonha I know you worked on the deprecation guide is this the latest up to date one I remember there had to be an issue or something to track what was deprecated?

hultonha commented:
> @hultonha I know you worked on the deprecation guide is this the latest up to date one I remember there had to be an issue or something to track what was deprecated?

Well spotted @pollend. @AMZN-Gene could you please refer to this document https://github.com/o3de/community/blob/main/guides/o3de-deprecation-guidelines.md and create a follow-up GHI (that is referenced at the callsite of the deprecation) to track its removal. This will help us actually track the deprecation and know when we can remove it. Thanks!

AMZN-Gene commented:
> > @hultonha I know you worked on the deprecation guide is this the latest up to date one I remember there had to be an issue or something to track what was deprecated?
> 
> Well spotted @pollend. @AMZN-Gene could you please refer to this document https://github.com/o3de/community/blob/main/guides/o3de-deprecation-guidelines.md and create a follow-up GHI (that is referenced at the callsite of the deprecation) to track its removal. This will help us actually track the deprecation and know when we can remove it. Thanks!

Nice! I added the GHI and updated this PR to include the GHI #12715

AMZN-Gene commented on Gems/ImGui/Code/Source/LYCommonMenu/ImGuiLYCommonMenu.cpp:
```diff
@@ -305,11 +305,14 @@ namespace ImGui
                 if (lvlSystem && ImGui::BeginMenu("Levels"))
                 {
-                    if (lvlSystem->IsLevelLoaded())
+                    if (AzFramework::LevelSystemLifecycleInterface::Get()->IsLevelLoaded())
```
Fixed!

lumberyard-employee-dm replied:
You could store this in a pointer for reuse later instead of making two calls to LevelSystemLifecycleInterface::`Get()`

hultonha commented:
> > > @hultonha I know you worked on the deprecation guide is this the latest up to date one I remember there had to be an issue or something to track what was deprecated?
> > 
> > 
> > Well spotted @pollend. @AMZN-Gene could you please refer to this document https://github.com/o3de/community/blob/main/guides/o3de-deprecation-guidelines.md and create a follow-up GHI (that is referenced at the callsite of the deprecation) to track its removal. This will help us actually track the deprecation and know when we can remove it. Thanks!
> 
> Nice! I added the GHI and updated this PR to include the GHI #12715

Awesome thanks @gene-walt! 🥳

kberg-amzn commented on Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.h:
```diff
@@ -20,34 +19,40 @@ namespace LegacyLevelSystem
         void AddListener(ILevelSystemListener* pListener) override;
         void RemoveListener(ILevelSystemListener* pListener) override;
 
         bool LoadLevel(const char* levelName) override;
```
We really just need this level system itself pulled out of legacy, or at least the interface, ILevelSystem, and have it registered as an AZ::Interface. This would be better than having a singleton level interface in the legacy codebase, and another ebus interface in azframework that we combine together in this manner. What we have here, we're going to increasingly need to move methods over to the LevelSystemLifecycle interface to expose it to the rest of the engine, and we'll likely end up needing to rename/refactor the LevelSystemLifecycle interface itself because the added functionality won't be lifecycle specific necessarily, resulting in more interface changes than strictly required.

AMZN-Gene replied:
Agreed! My next step was to pull this out entirely and potentially even pull CryCommon out of the other gems that need to use the level system. I’ll create a GHI for that task.

AMZN-Gene replied:
https://github.com/o3de/o3de/issues/12772
